### ISSUE TITLE
Add support for Proxy-Authorization

### ISSUE THREAD

**Issue by alippai** [pythongssapi/httpx-gssapi#3]
Similar to HTTP error 401 + Authorization headers sometimes not the server, but a proxy needs authorization.
It's using HTTP error 407 and Proxy-Authorization headers.

**Comment by aiudirog** [comment]
Is there any particular reason you can't use the standard proxy handling? Either through the HTTP[S]_PROXY environment variables or through the [routing interface](https://www.python-httpx.org/advanced/#routing)? It would be awkward and redundant to code the same functionality in every individual auth handler.

In theory, you should be able to do:

```python
client = httpx.Client(
    ...
    proxies={
        "http://": "http://USERNAME:PASSWORD@HOST:PORT",
        "https://": "https://USERNAME:PASSWORD@HOST:SSL_PORT",
    },
)
```
or 
```bash
export HTTP_PROXY=http://USERNAME:PASSWORD@HOST:PORT
export HTTPS_PROXY=https://USERNAME:PASSWORD@HOST:SSL_PORT

python my_code.py
```

**Comment by alippai** [comment]
I want to use password less kerberos interface (using keytab)

**Comment by aiudirog** [comment]
That doesn't have anything to do with proxy auth or much to do with this library really. GSSAPI authentication happens after you've acquired local Kerberos ticket through something like `kinit`. You can do that with a subprocess or some other Kerberos wrapper. 


Some simple subprocess code could be:

```python
subprocess.run(['kinit', 'USER@ADS', '-k', '-t', 'path/to/user.keytab'], check=True)
```

------

@frozencemetery Is there a way to kinit using a keytab with Python GSSAPI? I haven't explored that myself.

**Comment by alippai** [comment]
The Proxy-Authentication header can contain the kerberos ticket (to the proxy) similar to the Authentication header (to the service). Like you use gssapi to negotiate with the service instead of Basic auth upon HTTP error 401, it can be used to negotiate with the proxy instead of Basic auth upon HTTP error 407

**Comment by aiudirog** [comment]
Ohhh okay, sorry for the confusion, I understand now. You're trying to authenticate the proxy itself with GSSAPI. We can definitely do that, I'll just need to set up a test env for it.

**Comment by alippai** [comment]
Exactly, thanks for your patience and your responses. Next time I should start with a code snippet, sequence diag or some better illustration 🤔

**Comment by frozencemetery** [comment]
> @frozencemetery Is there a way to kinit using a keytab with Python GSSAPI? I haven't explored that myself.

(I recognize that the discussion has moved on here, but going to answer this one in case anyone finds this later.)

Yes, but which way depends on what you're trying to do :)

GSSAPI prefers to sit above the level of kinit.  But if you have a keytab, it can be specified using the KRB5_KEYTAB or KRB5_CLIENT_KEYTAB variables (see https://web.mit.edu/kerberos/krb5-latest/doc/basic/keytab_def.html) and then either calling `acquire_cred` or creating a security context as normal.  Alternately, the path can be given to `acquire_cred_from` as either the keytab or client_keytab argument in the store.

**Comment by aiudirog** [comment]
Yeah I was looking over that earlier. I've been mocking up a test environment using [proxy.py](https://pypi.org/project/proxy.py/) with a bare-bones GSSAPI auth plugin.

**Comment by aiudirog** [comment]
Just wanted to update that I am working on this, but I'm having a little trouble with the testing environment setup. I can get it to properly authenticate the proxy and connect to external websites, but connecting to the local webserver is causing the connection to be terminated. I'll figure it out, just wanted to provide an update.

**Comment by aiudirog** [comment]
Turns out, it was just a bug in proxy.py: https://github.com/abhinavsingh/proxy.py/issues/398. Setting the `Content-Length` header for our test cases solves the problem.

However, I've run into a new problem: I can't seem to determine the proxy host for generating the `SecurityContext` as it doesn't appear to be exposed as a part of the `Request`/`Response` class APIs. We may need to actually submit a PR to HTTPX to get this information in the response.

I've pushed my work so far to the `proxy-support` branch @ https://github.com/pythongssapi/httpx-gssapi/commit/82cca892dc85079a9d06a43bb3c379f15a2de1b1.

**Comment by enr0c** [comment]
Hey @aiudirog , do you have some news about this new feature? Would be great if it would work...

**Comment by aiudirog** [comment]
Hi @enr, sorry for the delayed response. I haven't followed up on this recently, but I'll see about looking into it again soon and see if anything has changed. You're also welcome to take a look if you want.

**Comment by aiudirog** [comment]
@enr0c I looked into this some more today and, if I'm reading this right, it appears proxies only support basic auth in HTTPX currently: https://github.com/encode/httpx/blob/0.28.1/httpx/_config.py#L207

The auth for the Client never even gets invoked against the proxy, so this will definitely require proper support from within HTTPX itself.

**Comment by aiudirog** [comment]
@v1gnesh This is a very hacky workaround that can work in scenarios where opportunistic auth is available and mutual auth is not required:
```python
import httpx
from httpx_gssapi import HTTPSPNEGOAuth

PROXY_URL = 'http://proxy.example:80/'


def make_proxy_auth(url: str) -> str:
    auth = HTTPSPNEGOAuth(opportunistic_auth=True)
    with httpx.Client() as client:
        req = client.build_request('GET', url)
        auth.set_auth_header(req)
        return req.headers['Authorization']


def make_proxy_conf(url: str) -> httpx.Proxy:
    return httpx.Proxy(
        url,
        headers={
            'Proxy-Authorization': make_proxy_auth(url),
        },
    )


def main():
    with httpx.Client(
        timeout=500,
        proxy=make_proxy_conf(PROXY_URL),
    ) as client:
        client.get("https://example.com/")
```

Basically you have `HTTPSPNEGOAuth` pre-generate an `Authorization` header and manually provide that to the `httpx.Proxy` as `Proxy-Authorization`, then pray it works.
